src/metabase/query_processor/middleware/update_used_cards.clj

```diff
@@ -0,0 +1,45 @@
+(ns metabase.query-processor.middleware.update-used-cards
+  (:require
+   [clojure.string :as str]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.util.log :as log]
+   #_{:clj-kondo/ignore [:discouraged-namespace]}
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *update-used-cards-async*
+  "Whether to update used cards asynchronously.
+  It's true by default and should only be changed for testing purposes"
+  true)
+
+(defn- update-used-cards!*
+  [used-cards]
+  (when (seq used-cards)
+    (let [f (fn []
+              (try
+                (log/debugf "%d cards were used during query execution: %s" (count used-cards) (str/join ";" (map :name used-cards)))
+                (t2/update! :model/Card :id [:in (map :id used-cards)] {:last_used_at :%now})
+                (catch Throwable e
+                  (log/error e "Error updating used cards"))))]
+      (if *update-used-cards-async*
+        (.submit clojure.lang.Agent/pooledExecutor ^Runnable f)
+        (f)))))
+
+(defn update-used-cards!
+  "Update last_used_at of all cards that were used during a query execution.
+
+  Including but not limited to cards used as:
+  - the source card for other queries
+  - definition for sandbox rules
+  - card references in native query
+  - dashcard on dashboard
+  - alert/pulses"
+  [qp]
+  (fn [query rff]
+    (letfn [(rff* [metadata]
+              ;; doing this async so it doesn't block execution
+              (update-used-cards!* (lib.metadata/cached-metadata (qp.store/metadata-provider) :metadata/card))
```

Comment:
This will return all the cards used by the query, but might return much more than that. The caching is not necessarily per query execution, or even per query.

To get a precise set, I'd wrap the metadata early in the process with a `composed-metadata-provider` that passes through all the calls except that card fetches are recorded. That feels like what you're going for with the `cached-metadata` approach, but without the risk that lots of unrelated cards happen to be cached already.

Reply:
> The caching is not necessarily per query execution, or even per query.

Oh, I'm surprised by this. Could you explain how that is possible? my understanding is that the cached provider is an atom and is initialized by calling `application-database-metadata-provider` with a `database-id.` 

I guess, in theory, someone could pass it with a provider that already has cached cards, but is it something we do today? and even in the future?

Reply:
I've looked into the implementation of `composed-metadata-provider`, and it seems like when fetching a `card`, it returns the result of the first provider that returns one.

so it seems like to make this work, we'll need to make sure that my custom provider has to be the first provider; it doesn't seem like a good guarantee to me. WDYT?